pkg/cmd/server/config.go

```diff
@@ -0,0 +1,254 @@
+	ClientConfig clientcmd.ClientConfig
+
+	CORSAllowedOrigins flagtypes.StringList
+}
+
+// GetMasterAddress checks for an unset master address and then attempts to use the first
+// public IPv4 non-loopback address registered on this host. It will also update the
+// EtcdAddr after if it was not provided.
+// TODO: make me IPv6 safe
+func (cfg config) GetMasterAddress() (*url.URL, error) {
+	if cfg.MasterAddr.Provided {
+		return cfg.MasterAddr.URL, nil
+	}
+
+	if cfg.IsStartMaster() {
+		// If the user specifies a bind address, and the master is not provided, use the bind port by default
+		port := cfg.MasterAddr.Port
+		if cfg.BindAddr.Provided {
+			port = cfg.BindAddr.Port
+		}
+
+		// If the user specifies a bind address, and the master is not provided, use the bind scheme by default
+		scheme := cfg.MasterAddr.URL.Scheme
+		if cfg.BindAddr.Provided {
+			scheme = cfg.BindAddr.URL.Scheme
+		}
+
+		// use the default ip address for the system
+		addr, err := util.DefaultLocalIP4()
+		if err != nil {
+			return nil, fmt.Errorf("Unable to find the public address of this master: %v", err)
+		}
+
+		masterAddr := scheme + "://" + net.JoinHostPort(addr.String(), strconv.Itoa(port))
+		return url.Parse(masterAddr)
+	}
+
+	// if we didn't specify and we aren't starting the master, read .kubeconfig to locate the master
+	// TODO client config currently doesn't let you override the defaults
+	// so it is defaulting to https://localhost:8443 for MasterAddr if
+	// it isn't set by --master or --kubeconfig
+	config, err := cfg.ClientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	return url.Parse(config.Host)
+}
+
+func (cfg config) GetMasterPublicAddress() (*url.URL, error) {
+	if cfg.MasterPublicAddr.Provided {
+		return cfg.MasterPublicAddr.URL, nil
+	}
+
+	return cfg.GetMasterAddress()
+}
+
+func (cfg config) GetEtcdAddress() (*url.URL, error) {
+	if cfg.EtcdAddr.Provided {
+		return cfg.EtcdAddr.URL, nil
+	}
+
+	// Etcd should be reachable on the same address that the master is (for simplicity)
+	masterAddr, err := cfg.GetMasterAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	etcdAddr := net.JoinHostPort(getHost(*masterAddr), strconv.Itoa(cfg.EtcdAddr.DefaultPort))
+	return url.Parse("http://" + etcdAddr)
+}
+
+func (cfg config) GetKubernetesAddress() (*url.URL, error) {
+	if cfg.KubernetesAddr.Provided {
+		return cfg.KubernetesAddr.URL, nil
+	}
+
+	return cfg.GetMasterAddress()
+}
+
+func (cfg config) GetKubernetesPublicAddress() (*url.URL, error) {
+	if cfg.KubernetesPublicAddr.Provided {
+		return cfg.KubernetesPublicAddr.URL, nil
+	}
+
+	return cfg.GetMasterPublicAddress()
+}
+
+func (cfg config) GetNodeList() []string {
+	nodeList := cfg.NodeList
+
+	if len(nodeList) == 1 && nodeList[0] == "127.0.0.1" {
+		nodeList[0] = cfg.Hostname
+	}
+	for i, s := range nodeList {
+		s = strings.ToLower(s)
+		nodeList[i] = s
+		glog.Infof("  Node: %s", s)
+	}
+
+	return nodeList
+}
+
+func (cfg config) IsStartNode() bool {
```

Comment:
Seems like the logic around how all the `IsStart*` things interact would be easier to understand calculated in config, then set as bools in the config:

```
if len(args) == 0 {
   cfg.StartMaster = true
   cfg.StartNode = true
}
if (len(args) == 1) && (args[0] == startMaster) {
   cfg.StartMaster = true
}
if (len(args) == 1) && (args[0] == startMaster) {
   cfg.StartNode = true
}
if cfg.StartMaster {
   cfg.StartEtcd = !cfg.EtcdAddr.Provided
   cfg.StartKube = !cfg.KubernetesAddr.Provided
}
```